File: templane-ts/tests/sidecar.test.ts
// Tests for SPEC §4.3 — sidecar mode.
import { test } from 'node:test';
import * as assert from 'node:assert/strict';
import * as fs from 'fs/promises';
import * as os from 'os';
import * as path from 'path';
import { parse, loadFromPath } from '../src/schema-parser';

async function mkTmpDir(prefix: string): Promise<string> {
  return await fs.mkdtemp(path.join(os.tmpdir(), prefix));
}

test('sidecar: body path recognized, not treated as field', () => {
  const yaml = 'body: ./email.jinja\nname:\n  type: string\n  required: true\n';
  const result = parse(yaml, 'sidecar-basic');
  assert.ok('schema' in result);
  assert.equal(result.bodyPath, './email.jinja');
  // body: key must NOT be treated as a field
  assert.ok(!('body' in result.schema.fields));
  assert.deepEqual(Object.keys(result.schema.fields), ['name']);
});

test('sidecar: explicit engine wins', () => {
  const yaml = 'body: ./t.hbs\nengine: jinja\nuser: {type: string, required: true}\n';
  const result = parse(yaml, 'engine-explicit');
  assert.ok('schema' in result);
  assert.equal(result.engine, 'jinja');
});

test('sidecar: engine inferred from extension', () => {
  const cases: [string, string][] = [
    ['./t.jinja',  'jinja'],
    ['./t.hbs',    'handlebars'],
    ['./t.ftl',    'freemarker'],
    ['./t.tmpl',   'gotemplate'],
    ['./t.md',     'markdown'],
    ['./t.html',   'html-raw'],
  ];
  for (const [bodyPath, expectedEngine] of cases) {
    const yaml = `body: ${bodyPath}\nname: {type: string, required: true}\n`;
    const result = parse(yaml, 'inferred');
    assert.ok('schema' in result, `unexpected error for ${bodyPath}`);
    assert.equal(result.engine, expectedEngine, `for ${bodyPath}`);
  }
});

test('sidecar: no extension, no engine inference', () => {
  const yaml = 'body: ./t\nname: {type: string, required: true}\n';
  const result = parse(yaml, 'no-ext');
  assert.ok('schema' in result);
  assert.equal(result.engine, undefined); // no extension → no inference, no error
});

test('sidecar: unknown engine rejected', () => {
  const yaml = 'body: ./t.jinja\nengine: mystery\nname: {type: string, required: true}\n';
  const result = parse(yaml, 'bad-engine');
  assert.ok('error' in result);
  assert.ok(result.error.includes('mystery'));
});

test('sidecar: absolute path rejected', () => {
  const yaml = 'body: /etc/passwd\nname: {type: string, required: true}\n';
  const result = parse(yaml, 'abs-path');
  assert.ok('error' in result);
  assert.ok(result.error.toLowerCase().includes('relative'));
});

test('sidecar: parent-escape path rejected', () => {
  const yaml = 'body: ../../../etc/passwd\nname: {type: string, required: true}\n';
  const result = parse(yaml, 'escape');
  assert.ok('error' in result);
});

test('sidecar: body: and --- conflict rejected', () => {
  const yaml = 'body: ./a.jinja\nname: {type: string, required: true}\n---\nHello\n';
  const result = parse(yaml, 'conflict');
  assert.ok('error' in result);
  assert.ok(result.error.toLowerCase().includes('both'));
});

test('embedded mode unchanged by 1.1', () => {
  const yaml = 'name:\n  type: string\n  required: true\n---\nHello {{ name }}!\n';
  const result = parse(yaml, 'embedded');
  assert.ok('schema' in result);
  assert.equal(result.bodyPath, undefined);
  assert.equal(result.engine, undefined);
  assert.equal(result.body, 'Hello {{ name }}!\n');
});

test('check-only mode (no body, no bodyPath)', () => {
  const yaml = 'name:\n  type: string\n  required: true\n';
  const result = parse(yaml, 'check-only');
  assert.ok('schema' in result);
  assert.equal(result.body, undefined);
  assert.equal(result.bodyPath, undefined);
});

test('loadFromPath resolves sidecar body file', async () => {
  const tmp = await mkTmpDir('templane-sidecar-');
  try {
    const bodyFile = path.join(tmp, 'greeting.jinja');
    await fs.writeFile(bodyFile, 'Hello {{ name }}!\n');
    const schemaFile = path.join(tmp, 'greeting.templane');
    await fs.writeFile(
      schemaFile,
      'body: ./greeting.jinja\nname: {type: string, required: true}\n',
    );

    const result = await loadFromPath(schemaFile);
    assert.ok('schema' in result);
    assert.equal(result.bodyPath, './greeting.jinja');
    assert.equal(result.body, 'Hello {{ name }}!\n');
    assert.equal(result.engine, 'jinja');
  } finally {
    await fs.rm(tmp, { recursive: true, force: true });
  }
});

test('loadFromPath: missing body file returns error', async () => {
  const tmp = await mkTmpDir('templane-sidecar-missing-');
  try {
    const schemaFile = path.join(tmp, 'broken.templane');
    await fs.writeFile(
      schemaFile,
      'body: ./nope.jinja\nname: {type: string, required: true}\n',
    );
    const result = await loadFromPath(schemaFile);
    assert.ok('error' in result);
    assert.ok(result.error.toLowerCase().includes('body file'));
  } finally {
    await fs.rm(tmp, { recursive: true, force: true });
  }
});

test('loadFromPath: embedded schema works like parse()', async () => {
  const tmp = await mkTmpDir('templane-sidecar-embedded-');
  try {
    const schemaFile = path.join(tmp, 'embedded.templane');
    await fs.writeFile(
      schemaFile,
      'name: {type: string, required: true}\n---\nHi {{ name }}\n',
    );
    const result = await loadFromPath(schemaFile);
    assert.ok('schema' in result);
    assert.equal(result.body, 'Hi {{ name }}\n');
    assert.equal(result.bodyPath, undefined);
  } finally {
    await fs.rm(tmp, { recursive: true, force: true });
  }
});

File: templane-go/core/sidecar_test.go
package core

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// Tests for SPEC §4.3 — sidecar mode.

func TestSidecarBodyPathRecognized(t *testing.T) {
	yaml := "body: ./email.jinja\nname:\n  type: string\n  required: true\n"
	r := ParseSchema(yaml, "sidecar-basic")
	if r.Error != "" {
		t.Fatalf("unexpected error: %s", r.Error)
	}
	if r.BodyPath == nil || *r.BodyPath != "./email.jinja" {
		t.Fatalf("body_path = %v, want ./email.jinja", r.BodyPath)
	}
	// body: key must NOT be treated as a field
	if _, ok := r.Schema.Fields["body"]; ok {
		t.Errorf("'body' must not be a field")
	}
	if len(r.Schema.Fields) != 1 {
		t.Errorf("expected 1 field, got %d", len(r.Schema.Fields))
	}
	if _, ok := r.Schema.Fields["name"]; !ok {
		t.Errorf("missing 'name' field")
	}
}

func TestSidecarEngineExplicit(t *testing.T) {
	yaml := "body: ./t.hbs\nengine: jinja\nuser: {type: string, required: true}\n"
	r := ParseSchema(yaml, "engine-explicit")
	if r.Error != "" {
		t.Fatalf("unexpected error: %s", r.Error)
	}
	// explicit engine wins even when extension disagrees
	if r.Engine == nil || *r.Engine != "jinja" {
		t.Errorf("engine = %v, want jinja", r.Engine)
	}
}

func TestSidecarEngineInferredFromExtension(t *testing.T) {
	cases := []struct {
		bodyPath string
		engine   string
	}{
		{"./t.jinja", "jinja"},
		{"./t.hbs", "handlebars"},
		{"./t.ftl", "freemarker"},
		{"./t.tmpl", "gotemplate"},
		{"./t.md", "markdown"},
		{"./t.html", "html-raw"},
	}
	for _, c := range cases {
		yaml := "body: " + c.bodyPath + "\nname: {type: string, required: true}\n"
		r := ParseSchema(yaml, "inferred")
		if r.Error != "" {
			t.Errorf("for %s: unexpected error %q", c.bodyPath, r.Error)
			continue
		}
		if r.Engine == nil || *r.Engine != c.engine {
			t.Errorf("for %s: engine = %v, want %s", c.bodyPath, r.Engine, c.engine)
		}
	}
}

func TestSidecarNoExtensionNoEngineInference(t *testing.T) {
	yaml := "body: ./t\nname: {type: string, required: true}\n"
	r := ParseSchema(yaml, "no-ext")
	if r.Error != "" {
		t.Fatalf("unexpected error: %s", r.Error)
	}
	// no extension → no inference (engine optional, no error)
	if r.Engine != nil {
		t.Errorf("engine = %v, want nil", *r.Engine)
	}
}

func TestSidecarUnknownEngineRejected(t *testing.T) {
	yaml := "body: ./t.jinja\nengine: mystery\nname: {type: string, required: true}\n"
	r := ParseSchema(yaml, "bad-engine")
	if r.Error == "" {
		t.Fatalf("expected error, got none")
	}
	if !strings.Contains(r.Error, "mystery") {
		t.Errorf("error %q does not mention 'mystery'", r.Error)
	}
}

func TestSidecarAbsolutePathRejected(t *testing.T) {
	yaml := "body: /etc/passwd\nname: {type: string, required: true}\n"
	r := ParseSchema(yaml, "abs-path")
	if r.Error == "" {
		t.Fatalf("expected error, got none")
	}
	if !strings.Contains(strings.ToLower(r.Error), "relative") {
		t.Errorf("error %q does not mention 'relative'", r.Error)
	}
}

func TestSidecarParentEscapeRejected(t *testing.T) {
	yaml := "body: ../../../etc/passwd\nname: {type: string, required: true}\n"
	r := ParseSchema(yaml, "escape")
	if r.Error == "" {
		t.Fatalf("expected error, got none")
	}
}

func TestSidecarAndSeparatorConflictRejected(t *testing.T) {
	yaml := "body: ./a.jinja\nname: {type: string, required: true}\n---\nHello\n"
	r := ParseSchema(yaml, "conflict")
	if r.Error == "" {
		t.Fatalf("expected error, got none")
	}
	if !strings.Contains(strings.ToLower(r.Error), "both") {
		t.Errorf("error %q does not mention 'both'", r.Error)
	}
}

func TestEmbeddedModeUnchangedBy1_1(t *testing.T) {
	// Backward compat: a 1.0 schema with --- separator still works.
	yaml := "name:\n  type: string\n  required: true\n---\nHello {{ name }}!\n"
	r := ParseSchema(yaml, "embedded")
	if r.Error != "" {
		t.Fatalf("unexpected error: %s", r.Error)
	}
	if r.BodyPath != nil {
		t.Errorf("body_path = %v, want nil", *r.BodyPath)
	}
	if r.Engine != nil {
		t.Errorf("engine = %v, want nil", *r.Engine)
	}
	if r.Body == nil || *r.Body != "Hello {{ name }}!\n" {
		t.Errorf("body = %v", r.Body)
	}
}

func TestCheckOnlyMode(t *testing.T) {
	// Schema with neither body: key nor --- → check-only, no body emitted.
	yaml := "name:\n  type: string\n  required: true\n"
	r := ParseSchema(yaml, "check-only")
	if r.Error != "" {
		t.Fatalf("unexpected error: %s", r.Error)
	}
	if r.Body != nil {
		t.Errorf("body = %v, want nil", *r.Body)
	}
	if r.BodyPath != nil {
		t.Errorf("body_path = %v, want nil", *r.BodyPath)
	}
}

func TestLoadSchemaFromPathResolvesSidecar(t *testing.T) {
	tmp := t.TempDir()
	bodyPath := filepath.Join(tmp, "greeting.jinja")
	if err := os.WriteFile(bodyPath, []byte("Hello {{ name }}!\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	schemaPath := filepath.Join(tmp, "greeting.templane")
	schemaSrc := "body: ./greeting.jinja\nname: {type: string, required: true}\n"
	if err := os.WriteFile(schemaPath, []byte(schemaSrc), 0o644); err != nil {
		t.Fatal(err)
	}

	r := LoadSchemaFromPath(schemaPath)
	if r.Error != "" {
		t.Fatalf("unexpected error: %s", r.Error)
	}
	if r.BodyPath == nil || *r.BodyPath != "./greeting.jinja" {
		t.Errorf("body_path = %v, want ./greeting.jinja", r.BodyPath)
	}
	if r.Body == nil || *r.Body != "Hello {{ name }}!\n" {
		t.Errorf("body = %v, want 'Hello {{ name }}!\\n'", r.Body)
	}
	if r.Engine == nil || *r.Engine != "jinja" {
		t.Errorf("engine = %v, want jinja", r.Engine)
	}
}

func TestLoadSchemaFromPathMissingBodyFile(t *testing.T) {
	tmp := t.TempDir()
	schemaPath := filepath.Join(tmp, "broken.templane")
	schemaSrc := "body: ./nope.jinja\nname: {type: string, required: true}\n"
	if err := os.WriteFile(schemaPath, []byte(schemaSrc), 0o644); err != nil {
		t.Fatal(err)
	}
	r := LoadSchemaFromPath(schemaPath)
	if r.Error == "" {
		t.Fatalf("expected error, got none")
	}
	if !strings.Contains(strings.ToLower(r.Error), "body file") {
		t.Errorf("error %q does not mention 'body file'", r.Error)
	}
}

func TestLoadSchemaFromPathEmbeddedMode(t *testing.T) {
	// load_from_path on an embedded schema should work the same as parse()
	tmp := t.TempDir()
	schemaPath := filepath.Join(tmp, "embedded.templane")
	schemaSrc := "name: {type: string, required: true}\n---\nHi {{ name }}\n"
	if err := os.WriteFile(schemaPath, []byte(schemaSrc), 0o644); err != nil {
		t.Fatal(err)
	}
	r := LoadSchemaFromPath(schemaPath)
	if r.Error != "" {
		t.Fatalf("unexpected error: %s", r.Error)
	}
	if r.Body == nil || *r.Body != "Hi {{ name }}\n" {
		t.Errorf("body = %v, want 'Hi {{ name }}\\n'", r.Body)
	}
	if r.BodyPath != nil {
		t.Errorf("body_path = %v, want nil", *r.BodyPath)
	}
}
